beater/beater.go

```diff
@@ -845,3 +913,55 @@ func newDropLogsBeatProcessor() (beat.ProcessorList, error) {
 		}),
 	})
 }
+
+type chanReloader struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	ch     chan reloadRequest
+}
+
+func newChanReloader() *chanReloader {
+	ctx, cancel := context.WithCancel(context.Background())
+	ch := make(chan reloadRequest)
+	return &chanReloader{ctx, cancel, ch}
+}
+
+type reloadRequest struct {
+	cfg    *reload.ConfigWithMeta
+	result chan<- error
+}
+
+func (r *chanReloader) Reload(cfg *reload.ConfigWithMeta) error {
+	result := make(chan error, 1)
+	select {
+	case <-r.ctx.Done():
+		return r.ctx.Err()
+	case r.ch <- reloadRequest{cfg: cfg, result: result}:
+	}
+	select {
+	case <-r.ctx.Done():
+		return r.ctx.Err()
+	case err := <-result:
+		return err
+	}
+}
+
+func (r *chanReloader) serve(ctx context.Context, reloader reload.Reloadable) error {
+	for {
+		select {
+		case <-r.ctx.Done():
+			return r.ctx.Err()
+		case <-ctx.Done():
+			return ctx.Err()
+		case req := <-r.ch:
+			err := reloader.Reload(req.cfg)
+			select {
+			case <-r.ctx.Done():
+				return r.ctx.Err()
+			case <-ctx.Done():
+				return ctx.Err()
+			case req.result <- err:
+			}
+		}
+	}
+}
```

Comment:
I find the `serve` and `Reload` functionality hard to follow; `Reload` is only called after receiving from the channel, but only `Reload` is sending to the channel. 
Would you mind adding some comment on this `chanReloader`?

Reply:
I've added some comments, please let me know if they help explain what's going on.